airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -471,6 +497,54 @@ def _checkpoint_mode(self) -> CheckpointMode:
         else:
             return CheckpointMode.FULL_REFRESH
 
+    @staticmethod
+    def _classify_stream(mappings_or_slices: Iterator[Optional[Union[Mapping[str, Any], StreamSlice]]]) -> StreamClassification:
+        """
+        This is a bit of a crazy solution, but also the only way we can detect certain attributes about the stream since Python
+        streams do not follow consistent implementation patterns. We care about the following two attributes:
+        - is_substream: Helps to incrementally release changes since substreams w/ parents are much more complicated. Also
+          helps de-risk the release of changes that might impact all connectors
+        - uses_legacy_slice_format: Since the checkpoint reader must manage a complex state object, we opted to have it always
+          use the structured StreamSlice object. However, this requires backwards compatibility with Python sources that only
+          support the legacy mapping object
+
+        Both attributes can eventually be deprecated once stream's define this method deleted once substreams have been implemented and
+        legacy connectors all adhere to the StreamSlice object.
+        """
+        if not mappings_or_slices:
+            raise ValueError("A stream should always have at least one slice")
+        try:
+            next_slice = next(mappings_or_slices)
+            if isinstance(next_slice, StreamSlice) and next_slice == StreamSlice(partition={}, cursor_slice={}):
+                is_legacy_format = False
+                slice_has_value = False
+            elif next_slice == {}:
+                is_legacy_format = True
+                slice_has_value = False
+            elif isinstance(next_slice, StreamSlice):
+                is_legacy_format = False
```

Comment:
Can we make this as `scenarios mapping` instead of reassigning it like this? 

```python
_STREAM_SCLIE_SCENARIOUS = {
        "slice_no_value": {"is_legacy_format": False, "slice_has_value": False},
        "slice_legacy_format": {"is_legacy_format": True, "slice_has_value": False},
        "slice_has_value": {"is_legacy_format": False, "slice_has_value": True},
    }
    
def _set_scenario(scenario) -> None:
    is_legacy_format = _STREAM_SCLIE_SCENARIOUS.get(scenario, {}).get("is_legacy_format")
    slice_has_value = _STREAM_SCLIE_SCENARIOUS.get(scenario, {}).get("slice_has_value")
```

Reply:
@bazarnov I might not be fully understanding this block of code. when we set/get the scenario and then convert it into a `StreamClassification` and return it. Is `_STREAM_SLICE_SCENARIOS` supposed to be a static map of a scenarios -> legacy format / has value? I'm not sure why we would need to set the value of the map?

I'm also not sure that this makes the code more or less readable personally. We need to use `slice_has_value` to continue to the second item of the iterator. And `slice_has_value` doesn't exactly map to the scenario we're trying to evaluate which is `has_multiple_slices`.

I think we're not quite thinking about this the same way